### DISCUSSION TITLE
Inheritance between templates

### DISCUSSION THREAD

**Discussion by fdaguin** [copier-org/copier#402] [Q&A] [answered]
In order to keep all my templates DRY, I'd like to define base templates which others will inherit from. 

Here is an example of what I'm trying to achieve:

![copier](https://user-images.githubusercontent.com/72013963/117504408-ac85ef00-af82-11eb-94f7-17e038a6fd56.png)

I want the child templates to have access to the answers of their parents.

How can I achieve this when using multiple `.copier-answers` files? Is there a way to read the answers of a specified `.copier-answers` file when cloning a template? 

Thanks !

**EDIT:** #79 seems close to what I'm trying to achieve.

**Comment by yajo** [comment, marked as answer]
Have you tried include? https://copier.readthedocs.io/en/stable/configuring/#include-other-yaml-files

**Reply by fdaguin** [reply]
## TL;DR

- if I'm not mistaken, the include feature doesn't provide mechanisms to read answers files
- when using nested includes, the source path used for the parents resolution is every time the path of the template called for the copy (see below for more details)
- to provide a definitive answer to #237, questions can indeed be imported from different sources
- it will be incredibly useful if parent templates could be cloned when their `copier.yml` file is sourced by a child → @Yajo, could it be possible?


## In depth

To provide additional context, here is the summary of my experiments.

### Setup

To test the include feature, here is the setup that I used:

```tree
.
└── templates
    ├── base
    │   └── copier.yml
    ├── ansible-base
    │   ├── ansible-base.copier.yml
    │   └── copier.yml
    └── ansible-role
        ├── ansible-role.copier.yml
        └── copier.yml
```

In `base/copier.yml`:

```yaml
_answers_file: .copier-answers.base.yml
# ...
# `base` specific questions.
# ...
```

In `ansible-base/ansible-base.copier.yml`:

```yaml
_answers_file: .copier-answers.ansible-base.yml
# ...
# `ansible-base` specific questions.
# ...
```

In `ansible-base/copier.yml`:

```yaml
!include ../base/copier.yml
---
!include ansible-base.copier.yml
```

In `ansible-role/ansible-role.copier.yml`:

```yaml
_answers_file: .copier-answers.ansible-role.yml
# ...
# `ansible-role` specific questions.
# ...
```

In `ansible-role/copier.yml`:

```yaml
!include ../ansible-base/copier.yml
---
!include ansible-role.copier.yml
```

### Experiments

#### Following the documentation ["Applying multiple templates to the same subproject"](https://github.com/copier-org/copier/blob/2a944bf73620f3c993155ff2754ee6218d093e62/docs/configuring.md#applying-multiple-templates-to-the-same-subproject)

I tried to clone each templates, in the following order:

1. `base`
2. `ansible-base`
3. `ansible-role`

In action, I first clone the `base` template..

```bash
copier copy base/ temp/
```

..which produces `temp/.copier-answers.base.yml`.

However, when I clone the `ansible-base` template..

```bash
copier copy ansible-base/ temp/
```

⚠️ ..all the questions from `base` are asked again.
⚠️ ..the answers of `temp/.copier-answers.base.yml` are not read because I don't know how to do it when configuring the `ansible-base` template.
💡 In `ansible-base/copier.yml`, it could be useful if I was able to include the answers like so:

```yaml
!include ../base/copier.yml
---
!include "{{ copier.dst_path }}/.copier-answers.base.yml"
---
!include ansible-base.copier.yml
```

#### Only applying the *deepest* child template

I still wanted to try if applying the template `ansible-role` will work in a true inheritance fashion. In other word, I was trying to see if all the parent templates will be cloned when only invoking their child.

##### First try

In the `templates` directory, when running..

```bash
copier copy ansible-role/ temp/
```

..I get the following error:

```text
FileNotFoundError: [Errno 2] No such file or directory: '/templates/ansible-role/ansible-base.copier.yml'
```

`ansible-base/ansible-base.copier.yml` is trying to be resolved from the `ansible-role` folder.

⚠️ Quick note when using nested includes: the source path used for the parents resolution is **every time** the path of the template called for the copy.

##### Second try

I slightly tweaked my setup.

In `ansible-base/copier.yml`:

```yaml
_answers_file: .copier-answers.ansible-base.yml
```

The `ansible-base/ansible-base.copier.yml` file is no longer useful.

In `ansible-role/copier.yml`:

```yaml
!include ../base/copier.yml
---
!include ../ansible-base/copier.yml
---
!include ansible-role.copier.yml
```

As expected, when running..

```bash
copier copy ansible-role/ temp/
```

- ..all the questions defined in the parent templates are asked
- ..all the answers are stored in the file `temp/.copier-answers.ansible-role.yml` (including answers of the parents questions)
- ..only the files defined in the template `ansible-role` are copied in the destination folder

**Reply by yajo** [reply]
I kinda think you're misunderstanding some features. Some things you need to understand to not become crazy:

1. `!include` is for config files, not answers. You cannot include an answers file. Also, it combines several questions files, flattened in one. That's why questions are asked twice if you include the same config in 2 templates.

2. You can apply more than one template to a project, if each answers file has its own name. 1 answers file always relates to 1 template. In this case, each template should be atomic and care only about its own files. 

3. You can include shared configs or files with git submodules. 
4. Your template only needs to remember answers if you plan to support upgrades. Some templates like ansible-role seem like bootstrap-only use cases.
5. Templates only support updates if they are stored in versioned git repos.

Possibly knowing these facts will help you get what you want. 😉

**Reply by fdaguin** [reply]
Thanks a lot for this in-depth answer, this is much clearer now!

I will stick with git capabilities for the vendoring part.
